sdk/storage/azure-storage-common-data-movement/src/main/java/com/azure/storage/common/data/movement/PathScanner.java

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.common.data.movement;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.*;
+import java.util.stream.BaseStream;
+import java.util.stream.Stream;
+
+public class PathScanner {
+    private final String basePath;
+    private final boolean isDir;
+
+    public PathScanner(String path, boolean isDir) {
+        basePath = path;
+        this.isDir = isDir;
+    }
+
+    public Flux<String> scan(boolean skipSubdirectories) {
+        // Iterate through the path if it's a directory
+        if (isDir) {
+            // Set up Mono for recursing through directories
+            Mono<Path> basePath = Mono.just(Paths.get(this.basePath));
+
+            // Return a Flux constructed from expanding path
+            return basePath.expand(path ->
+                Flux.using(() -> Files.list(path),
+                    Flux::fromStream,
+                    BaseStream::close)
+                    .onErrorResume((e) -> {
+                        // Ignore errors resulting from trying to enumerate contents of
+                        // a file as if it were a directory, but log other errors
+                        if (!(e instanceof NotDirectoryException))
+                            System.err.println(e);
+
+                        // If set to skip subdirectories, continue processing; else,
+                        // pass an error which will stop the Flux stream
+                        if (skipSubdirectories)
+                            return Mono.empty();
```

Comment:
how this can happen?

Reply:
If we lack permission to read/enumerate a folder, it'll throw AccessDeniedException, so I have the argument here to let a user decide whether or not to cut all enumeration at that point or continue. (Should log either way)

Reply:
is there a way to check if we can enumerate before attempting to do so ?

Reply:
Yes, we can check ACL/file permissions depending on the platform. I'll add that in.

Reply:
Actually, I think the reason I didn't put a check beforehand here is because of issues that may arise from permissions changing between the read permission check and the actual read/enumeration, however unlikely. There were similar checks that were removed for .NET due to the same concern; @amnguye what do you think? Same concept here?